src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -195,6 +197,15 @@ var ReactCompositeComponentMixin = {
           Component.displayName || Component.name || 'Component'
         );
       }
+
+      propsMutated = !shallowEqual(inst.props, publicProps);
```

Comment:
That was my initial reaction too, but if I understand this code correctly, that check is already dev only.